awx/api/generics.py

```diff
@@ -712,3 +716,159 @@ def get_queryset(self):
         for r in roles:
             ancestors.update(set(r.ancestors.all()))
         return User.objects.filter(roles__in=list(ancestors)).distinct()
+
+
+def trigger_delayed_deep_copy(*args, **kwargs):
+    from awx.main.tasks import deep_copy_model_obj
+    connection.on_commit(lambda: deep_copy_model_obj.delay(*args, **kwargs))
+
+
+class CopyAPIView(GenericAPIView):
+
+    serializer_class = CopySerializer
+    copy_return_serializer_class = None
+    new_in_330 = True
+    new_in_api_v2 = True
+
+    IS_COPY_VIEW = True
+
+    def _get_copy_return_serializer(self, *args, **kwargs):
+        if not self.copy_return_serializer_class:
+            return self.get_serializer(*args, **kwargs)
+        serializer_class_store = self.serializer_class
+        self.serializer_class = self.copy_return_serializer_class
+        ret = self.get_serializer(*args, **kwargs)
+        self.serializer_class = serializer_class_store
+        return ret
+
+    @staticmethod
+    def _decrypt_model_field_if_needed(obj, field_name, field_val):
+        if isinstance(field_val, dict):
+            for sub_field in field_val:
+                if isinstance(sub_field, (str, unicode)) \
+                        and isinstance(field_val[sub_field], (str, unicode)):
+                    try:
+                        field_val[sub_field] = decrypt_field(obj, field_name, sub_field)
+                    except InvalidToken:
+                        # Catching the corner case with v1 credential fields
+                        field_val[sub_field] = decrypt_field(obj, sub_field)
+        elif isinstance(field_val, (str, unicode)):
+            field_val = decrypt_field(obj, field_name)
+        return field_val
```

Comment:
I can't easily tell what is happening here. It looks like the first block is for decrypting values inside of dictionaries (which hints at `extra_vars`) but the 2nd block looks like it's decrypting password fields. Firstly, I don't think that the same call pattern of `decrypt_field` will work for both. Secondly, I don't think you actually need to decrypt anything for copying encrypted extra_vars anyway.

It would be much more clear if you could do a more direct check to see if the field is marked as a password field.

Reply:
The call pattern of `decrypt_field` are different for the 2 blocks; 

`decrypt_field` function, on the other hand, is able to distinguish between encrypted and non-encrypted fields by checking `$encrypted$` prefix of string value. The reason I don't rely on encryption declaration is because there aren't any universal way of declaration out there, other than checking by `decrypt_field`.

Reply:
Question: the block of `if isinstance(field_val, dict):`, what model and field is that intended for?

My suspected answer: I imagine that it is for `inputs` on the Credential model.

I looked through the tests, and didn't see any example with a survey_spec with an encrypted default. If I read this right, that field (JT.survey_spec and WFJT.survey_spec) will get wrapped up in your logic here, and I suspect it won't work right.

The defaults in the survey spect are encrypted via a different key than other models. `decrypt_field` uses the model's primary key in the encryption key. If you do that with the survey values, the result will be a failure of decryption.

This will also apply to WFJT nodes and schedules, which now host encrypted password survey question answers.

In all cases of survey-related dict fields, they can be copied as-is, and no decryption/encryption cycle is needed. So this needs to be modified somehow to behavior in that way.

Reply:
I see what you mean. But passwords in survey specs are not affected by re-encryption, because those password fields are not on the first JSON level:
```
{
    "description": "",
    "name": "",
    "spec": [
        {
            "question_description": "",
            "min": 0,
            "default": "$encrypted$",
            "max": 32,
            "required": true,
            "choices": "",
            "new_question": true,
            "variable": "hoho",
            "question_name": "Hey",
            "type": "password"
        }
    ]
}
```

But you are right, it is safer to also check `PASSWORD_FIELDS` to prevent any accidental false re-encryption.

Reply:
Oh no, `PASSWORD_FIELDS` cannot be relied upon because there are custom password fields not listed there, like notification template configrations: https://github.com/ansible/awx/blob/devel/awx/main/models/notifications.py#L114.